### ISSUE TITLE
Better document how to migrate borrowing `extract()` calls

### ISSUE THREAD

**Issue by djc** [PyO3/pyo3#3958]
### Bug Description

To get a taste for the 0.21.0 migration I tried to migrate [instant-segment](https://github.com/instant-labs/instant-segment) over. If I manually enable `gil-refs` (I was a little surprised that wasn't default enabled), it compiles fine. However, if I remove `gil-refs` I get some compilation failures. However, reading the [migration guide](https://pyo3.rs/v0.21.0-beta.0/migration#from-020-to-021) didn't help me find a solution for this easily. Searching for `FromPyObject` finds some prose on that, mentioning the `extract_bound()` method but it didn't seem to help for this.

As discussed here: https://hachyderm.io/@davidhewitt@fosstodon.org/112077791453594471. Even with the additional hints on `FromPyObjectBound`, I didn't really figure out how to apply that here.

```rust
error[E0277]: the trait bound `&str: PyClass` is not satisfied
   --> instant-segment-py/src/lib.rs:38:45
    |
38  |                 let key = item.get_item(0)?.extract::<&str>()?;
    |                                             ^^^^^^^ the trait `PyClass` is not implemented for `&str`
    |
    = help: the following other types implement trait `PyClass`:
              Segmenter
              Search
    = note: required for `&str` to implement `FromPyObject<'_>`
note: required by a bound in `PyAny::extract`
   --> /Users/djc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-0.21.0-beta.0/src/types/any.rs:806:12
    |
804 |     pub fn extract<'py, D>(&'py self) -> PyResult<D>
    |            ------- required by a bound in this associated function
805 |     where
806 |         D: FromPyObject<'py>,
    |            ^^^^^^^^^^^^^^^^^ required by this bound in `PyAny::extract`
```

### Steps to Reproduce

1. `git clone https://github.com/instant-labs/instant-segment`
2. Update `instant-segment-py/Cargo.toml` to depend on pyo3 `=0.21.0-beta.0`
3. Run `cargo check`

### Your operating system and version

macOS 14.4

### Your Python version (`python --version`)

Python 3.9.6 (default, Feb  3 2024, 15:58:27)

### Your Rust version (`rustc --version`)

rustc 1.76.0 (07dca489a 2024-02-04)

### Your PyO3 version

0.21.0-beta.0

### How did you install python? Did you use a virtualenv?

System Python 3.

### Additional Info

FWIW, I've always found the `.parse()` form of `FromStr` a little too magic as a reader, and I usually use `Type::from_str()` instead. It would be nice if that was used more often/was ergonomic enough for these kinds of things (though chaining like in this particular context maybe too hard?).

**Comment by Icxolu** [comment]
I think this particular error is caused, because you still use `gil-ref`s as arguments.

```rust
fn new(unigrams: &PyIterator, bigrams: &PyIterator) -> PyResult<Self> {
```

These use the `extract` method on `PyAny`, which is bound by `FromPyObject`. If the `gil-ref` feature is disabled `&str` does _not_ implement `FromPyObject` anymore, instead `FromPyObjectBound` is implemented. This allows the extraction of `&str` via `PyAnyMethods::extract`, which is bounded by `FromPyObjectBound`. If you change the arguments to use `Bound` this starts working again (with introduction of temporaries to make the lifetimes work).
```rust
fn new(unigrams: Bound<'_, PyIterator>, bigrams: Bound<'_, PyIterator>) -> PyResult<Self> {
```

We are planning to add deprecation warnings to `gil-ref` arguments. That would maybe help a little bit. I'm not sure if it's possible to make this extraction still work with `gil-refs` disabled, in any case we should add a bit more context and maybe even an example to the migration guide.

**Comment by davidhewitt** [comment]
Thanks @djc for following up here!

On top of what @Icxolu says about the cases where maybe the deprecation warnings still missing has not helped you, I think the the example code you write is a perfect one which we should cover in the migration notes:

```rust
let key = item.get_item(0)?.extract::<&str>()?;
```

This pattern is something which is impossible without `gil-refs` because the `.get_item()` call is returning an owned Python reference. The solution we can recommend in the migration guide is to split this up:

```rust
let key = item.get_item(0)?
let key = key.extract::<&str>()?;
```

---

> (I was a little surprised that wasn't default enabled)

I agree that not making `gil-refs` a default feature is a little painful for users on first upgrade. However, I also think it is correct to not do this.

The problem I see with adding `gil-refs` to the default features is that then for users to migrate off GIL Refs they have to put `default-features = false` in their Cargo.toml. Once they have done this once, they probably forget they have done this, so if we ever have to do a deprecation-with-a-default-feature migration again then those users are immediately broken.

In summary, I saw default feature as a one-time trick rather than something which is actually a repeatable migration tool.

---

Thinking about the migration pathway, I am tempted to add a `gil-refs-migration` feature which keeps the `FromPyObject` implementations around but shows all the deprecation warnings. (It's a kind of halfway point between the current choice of `gil-refs` and no `gil-refs`. The `gil-refs` trait would automatically enable it.)

The nice thing about this would be that the migration from 0.20 would then be
- Enable `gil-refs`, fix the rest of 0.21 changes
- Switch to `gil-refs-migration`, fix the warnings
- Remove `gil-refs-migration`, fix the final breaking errors

I think that's smoother than just asking users to turn off the `gil-refs` feature and try to figure out how to fix all the deprecation warnings while also being blocked by compile errors.

**Comment by djc** [comment]
> Thinking about the migration pathway, I am tempted to add a `gil-refs-migration` feature which keeps the `FromPyObject` implementations around but shows all the deprecation warnings. (It's a kind of halfway point between the current choice of `gil-refs` and no `gil-refs`. The `gil-refs` trait would automatically enable it.)

Yes, this sounds great.

**Comment by davidhewitt** [comment]
Will reopen until I get `gil-refs-migration` pushed too (hopefully later tonight can get the PR up).

**Comment by davidhewitt** [comment]
In the end based on feedback in #3962 I think the solution in #3959 should actually be sufficient. That should cause the break in lifetime semantics to only apply when changing code over to the Bound API, which is a much better experience.

I'll also open a PR now with some additional docs detail.
